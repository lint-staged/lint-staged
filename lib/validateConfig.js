/** @typedef {import('./index').Logger} Logger */

import { inspect } from 'node:util'

import debug from 'debug'

import { configurationError, failedToParseConfig } from './messages.js'
import { ConfigEmptyError, ConfigFormatError } from './symbols.js'
import { validateBraces } from './validateBraces.js'

const debugLog = debug('lint-staged:validateConfig')

export const validateConfigLogic = (config, configPath, logger) => {
  debugLog('Validating config from `%s`...', configPath)

  if (!config || (typeof config !== 'object' && typeof config !== 'function')) {
    throw ConfigFormatError
  }

  /**
   * Function configurations receive all staged files as their argument.
   * They are not further validated here to make sure the function gets
   * evaluated only once.
   *
   * @see getSpawnedTasks
   */
  if (typeof config === 'function') {
    return { '*': config }
  }

  if (Object.entries(config).length === 0) {
    throw ConfigEmptyError
  }

  const errors = []

  /**
   * Create a new validated config because the keys (patterns) might change.
   * Since the Object.reduce method already loops through each entry in the config,
   * it can be used for validating the values at the same time.
   */
  const validatedConfig = Object.entries(config).reduce((collection, [pattern, task]) => {
    if (Array.isArray(task)) {
      /** Array with invalid values */
      if (task.some((item) => typeof item !== 'string' && typeof item !== 'function')) {
        errors.push(
          configurationError(pattern, 'Should be an array of strings or functions.', task)
        )
      }
    } else if (typeof task === 'object') {
      /** Invalid function task */
      if (typeof task.title !== 'string' || typeof task.task !== 'function') {
        errors.push(
          configurationError(
            pattern,
            'Function task should contain `title` and `task` fields, where `title` should be a string and `task` should be a function.',
            task
          )
        )
      }
    } else if (typeof task !== 'string' && typeof task !== 'function') {
      /** Singular invalid value */
      errors.push(
        configurationError(
          pattern,
          'Should be a string, a function, an object or an array of strings and functions.',
          task
        )
      )
    }

    /**
     * A typical configuration error is using invalid brace expansion, like `*.{js}`.
     * These are automatically fixed and warned about.
     */
    const fixedPattern = validateBraces(pattern, logger)

    return Object.assign(collection, { [fixedPattern]: task })
  }, {})

  if (errors.length) {
    const message = errors.join('\n\n')

    logger.error(failedToParseConfig(configPath, message))

    throw new Error(message)
  }

  debugLog('Validated config from `%s`:', configPath)
  debugLog(inspect(config, { compact: false }))

  return validatedConfig
}

/**
 * Runs config validation. Throws error if the config is not valid.
 * @param {Object} config
 * @param {string} configPath
 * @param {Logger} logger
 * @returns {Object} config
 */
export const validateConfig = (config, configPath, logger) => {
  try {
    return validateConfigLogic(config, configPath, logger)
  } catch (error) {
    logger.error(failedToParseConfig(configPath, error))
    throw error
  }
}
