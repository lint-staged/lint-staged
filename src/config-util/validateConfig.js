'use strict'

const chalk = require('chalk')
const format = require('stringify-object')
const jestValidate = require('jest-validate')
const unknownOptionWarning = require('jest-validate/build/warnings').unknownOptionWarning
const isGlob = require('is-glob')
const exampleConfig = require('./defaultConfig').exampleConfig

const validate = jestValidate.validate
const logValidationWarning = jestValidate.logValidationWarning

/**
 * Custom jest-validate reporter for unknown options
 * @param config
 * @param example
 * @param option
 * @param options
 * @returns {void}
 */
function unknownValidationReporter(config, example, option, options) {
  /**
   * If the unkonwn property is a glob this is probably
   * a typical mistake of mixing simple and advanced configs
   */
  if (isGlob(option)) {
    const message = `  Unknown option ${chalk.bold(`"${option}"`)} with value ${chalk.bold(
      format(config[option], { inlineCharacterLimit: Number.POSITIVE_INFINITY })
    )} was found in the config root.

  You are probably trying to mix simple and advanced config formats. Adding

  ${chalk.bold(`"linters": {
    "${option}": ${JSON.stringify(config[option])}
  }`)}

  will fix it and remove this message.`

    const comment = options.comment
    const name = (options.title && options.title.warning) || 'WARNING'
    return logValidationWarning(name, message, comment)
  }
  // If it is not glob pattern, use default jest-validate reporter
  return unknownOptionWarning(config, example, option, options)
}

/**
 * Runs config validation. Throws error if the config is not valid.
 * @param config {Object}
 * @returns config {Object}
 */
module.exports = function validateConfig(config) {
  const validation = validate(config, {
    exampleConfig,
    unknown: unknownValidationReporter,
    comment:
      'Please refer to https://github.com/okonet/lint-staged#configuration for more information...'
  })

  if (!validation.isValid) {
    throw new Error('lint-staged config is invalid... Aborting.')
  }

  return config
}
