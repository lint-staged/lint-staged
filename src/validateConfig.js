/* eslint no-prototype-builtins: 0 */

'use strict'

const chalk = require('chalk')
const format = require('stringify-object')

const debug = require('debug')('lint-staged:cfg')

const TEST_DEPRECATED_KEYS = new Map([
  ['concurrent', key => typeof key === 'boolean'],
  ['chunkSize', key => typeof key === 'number'],
  ['globOptions', key => typeof key === 'object'],
  ['linters', key => typeof key === 'object'],
  ['ignore', key => Array.isArray(key)],
  ['subTaskConcurrency', key => typeof key === 'number'],
  ['renderer', key => typeof key === 'string'],
  ['relative', key => typeof key === 'boolean']
])

const formatError = helpMsg => `● Validation Error:

  ${helpMsg}

Please refer to https://github.com/okonet/lint-staged#configuration for more information...`

const createError = (opt, helpMsg, value) =>
  formatError(`Invalid value for '${chalk.bold(opt)}'.

  ${helpMsg}.
 
  Configured value is: ${chalk.bold(
    format(value, { inlineCharacterLimit: Number.POSITIVE_INFINITY })
  )}`)

/**
 * Runs config validation. Throws error if the config is not valid.
 * @param config {Object}
 * @returns config {Object}
 */
module.exports = function validateConfig(config) {
  debug('Validating config')

  const errors = []

  if (!config || typeof config !== 'object') {
    errors.push('Configuration should be an object!')
  } else {
    const entries = Object.entries(config)

    if (entries.length === 0) {
      errors.push('Configuration should not be empty!')
    }

    entries.forEach(([pattern, task]) => {
      if (TEST_DEPRECATED_KEYS.has(pattern)) {
        const testFn = TEST_DEPRECATED_KEYS.get(pattern)
        if (testFn(task)) {
          errors.push(
            createError(
              pattern,
              'Advanced configuration has been deprecated. For more info, please visit: https://github.com/okonet/lint-staged',
              task
            )
          )
        }
      }

      if (
        (!Array.isArray(task) ||
          task.some(item => typeof item !== 'string' && typeof item !== 'function')) &&
        typeof task !== 'string' &&
        typeof task !== 'function'
      ) {
        errors.push(
          createError(
            pattern,
            'Should be a string, a function, or an array of strings and functions',
            task
          )
        )
      }

      entries.forEach(([, task]) => {
        if (typeof task !== 'function') return
        const resolved = task(['[filename]'])
        if (typeof resolved === 'string') return
        if (!Array.isArray(resolved) || resolved.some(subtask => typeof subtask !== 'string')) {
          errors.push(
            createError(
              task,
              'Function task should return a string or an array of strings',
              resolved
            )
          )
        }
      })
    })
  }

  if (errors.length) {
    throw new Error(errors.join('\n'))
  }

  return config
}
