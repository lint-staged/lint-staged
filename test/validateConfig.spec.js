import makeConsoleMock from 'consolemock'

import validateConfig from '../src/validateConfig'

describe('validateConfig', () => {
  const originalConsole = global.console
  beforeAll(() => {
    global.console = makeConsoleMock()
  })

  beforeEach(() => {
    global.console.clearHistory()
  })

  afterAll(() => {
    global.console = originalConsole
  })

  it('should throw and should print validation errors for invalid config 1', () => {
    const invalidConfig = 'test'
    expect(() => validateConfig(invalidConfig)).toThrowErrorMatchingSnapshot()
  })

  it('should throw and should print validation errors for invalid config', () => {
    const invalidConfig = {
      foo: false
    }
    expect(() => validateConfig(invalidConfig)).toThrowErrorMatchingSnapshot()
  })

  it('should not throw and should print nothing for valid config', () => {
    const validSimpleConfig = {
      '*.js': ['eslint --fix', 'git add']
    }
    expect(() => validateConfig(validSimpleConfig)).not.toThrow()
    expect(console.printHistory()).toMatchSnapshot()
  })

  it('should not throw and should print nothing for function task', () => {
    expect(() =>
      validateConfig({
        '*.js': filenames => {
          const files = filenames.join(' ')
          return `eslint --fix ${files} && git add ${files}`
        },
        '*.css': [filenames => filenames.map(filename => `eslint --fix ${filename}`)]
      })
    ).not.toThrow()
    expect(console.printHistory()).toMatchSnapshot()
  })

  it('should throw when function task returns incorrect values', () => {
    const invalidConfig = {
      '*.js': filenames => filenames.map(file => [`eslint --fix ${file}`, `git add ${file}`])
    }
    expect(() => validateConfig(invalidConfig)).toThrowErrorMatchingSnapshot()
  })

  it('should throw when detecting deprecated advanced configuration', () => {
    const advancedConfig = {
      chunkSize: 10,
      concurrent: false,
      globOptions: { matchBase: false },
      ignore: ['test.js'],
      linters: {
        '*.js': ['eslint']
      },
      relative: true,
      renderer: 'silent',
      subTaskConcurrency: 10
    }

    expect(() => validateConfig(advancedConfig)).toThrowErrorMatchingSnapshot()
    expect(console.printHistory()).toMatchSnapshot()
  })

  it('should not throw when config contains deprecated key but with valid task', () => {
    const stillValidConfig = {
      concurrent: 'my command'
    }
    expect(() => validateConfig(stillValidConfig)).not.toThrow()
    expect(console.printHistory()).toMatchSnapshot()
  })
})
