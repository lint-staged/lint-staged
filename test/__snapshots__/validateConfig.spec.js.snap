// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateConfig should not throw and should print nothing for function task 1`] = `""`;

exports[`validateConfig should not throw and should print nothing for valid config 1`] = `""`;

exports[`validateConfig should not throw when config contains deprecated key but with valid task 1`] = `""`;

exports[`validateConfig should throw and should print validation errors for invalid config 1`] = `
"● Validation Error:

  Invalid value for 'foo'.

  Should be a string, a function, or an array of strings and functions.
 
  Configured value is: false

Please refer to https://github.com/okonet/lint-staged#configuration for more information..."
`;

exports[`validateConfig should throw and should print validation errors for invalid config 1 1`] = `"Configuration should be an object!"`;

exports[`validateConfig should throw when detecting deprecated advanced configuration 1`] = `
"● Validation Error:

  Invalid value for 'chunkSize'.

  Advanced configuration has been deprecated. For more info, please visit: https://github.com/okonet/lint-staged.
 
  Configured value is: 10

Please refer to https://github.com/okonet/lint-staged#configuration for more information...
● Validation Error:

  Invalid value for 'chunkSize'.

  Should be a string, a function, or an array of strings and functions.
 
  Configured value is: 10

Please refer to https://github.com/okonet/lint-staged#configuration for more information...
● Validation Error:

  Invalid value for 'concurrent'.

  Advanced configuration has been deprecated. For more info, please visit: https://github.com/okonet/lint-staged.
 
  Configured value is: false

Please refer to https://github.com/okonet/lint-staged#configuration for more information...
● Validation Error:

  Invalid value for 'concurrent'.

  Should be a string, a function, or an array of strings and functions.
 
  Configured value is: false

Please refer to https://github.com/okonet/lint-staged#configuration for more information...
● Validation Error:

  Invalid value for 'globOptions'.

  Advanced configuration has been deprecated. For more info, please visit: https://github.com/okonet/lint-staged.
 
  Configured value is: {matchBase: false}

Please refer to https://github.com/okonet/lint-staged#configuration for more information...
● Validation Error:

  Invalid value for 'globOptions'.

  Should be a string, a function, or an array of strings and functions.
 
  Configured value is: {matchBase: false}

Please refer to https://github.com/okonet/lint-staged#configuration for more information...
● Validation Error:

  Invalid value for 'ignore'.

  Advanced configuration has been deprecated. For more info, please visit: https://github.com/okonet/lint-staged.
 
  Configured value is: ['test.js']

Please refer to https://github.com/okonet/lint-staged#configuration for more information...
● Validation Error:

  Invalid value for 'linters'.

  Advanced configuration has been deprecated. For more info, please visit: https://github.com/okonet/lint-staged.
 
  Configured value is: {'*.js': ['eslint']}

Please refer to https://github.com/okonet/lint-staged#configuration for more information...
● Validation Error:

  Invalid value for 'linters'.

  Should be a string, a function, or an array of strings and functions.
 
  Configured value is: {'*.js': ['eslint']}

Please refer to https://github.com/okonet/lint-staged#configuration for more information...
● Validation Error:

  Invalid value for 'relative'.

  Advanced configuration has been deprecated. For more info, please visit: https://github.com/okonet/lint-staged.
 
  Configured value is: true

Please refer to https://github.com/okonet/lint-staged#configuration for more information...
● Validation Error:

  Invalid value for 'relative'.

  Should be a string, a function, or an array of strings and functions.
 
  Configured value is: true

Please refer to https://github.com/okonet/lint-staged#configuration for more information...
● Validation Error:

  Invalid value for 'renderer'.

  Advanced configuration has been deprecated. For more info, please visit: https://github.com/okonet/lint-staged.
 
  Configured value is: 'silent'

Please refer to https://github.com/okonet/lint-staged#configuration for more information...
● Validation Error:

  Invalid value for 'subTaskConcurrency'.

  Advanced configuration has been deprecated. For more info, please visit: https://github.com/okonet/lint-staged.
 
  Configured value is: 10

Please refer to https://github.com/okonet/lint-staged#configuration for more information...
● Validation Error:

  Invalid value for 'subTaskConcurrency'.

  Should be a string, a function, or an array of strings and functions.
 
  Configured value is: 10

Please refer to https://github.com/okonet/lint-staged#configuration for more information..."
`;

exports[`validateConfig should throw when detecting deprecated advanced configuration 2`] = `""`;

exports[`validateConfig should throw when function task returns incorrect values 1`] = `
"● Validation Error:

  Invalid value for 'filenames => filenames.map(file => [\`eslint --fix \${file}\`, \`git add \${file}\`])'.

  Function task should return a string or an array of strings.
 
  Configured value is: [['eslint --fix [filename]', 'git add [filename]']]

Please refer to https://github.com/okonet/lint-staged#configuration for more information..."
`;
